File: src/main/java/corecode/WordCount.java
package corecode;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class WordCount {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setMaster("local");
        conf.setAppName("java_wc");
        JavaSparkContext sc = new JavaSparkContext(conf);
      JavaRDD<String> lines = sc.textFile("./data/words");


//        //filter
//        JavaRDD<String> filter = lines.filter(new Function<String, Boolean>() {
//            @Override
//            public Boolean call(String s) throws Exception {
//                return "hello hbase".equals(s);
//            }
//        });
//        filter.foreach(new VoidFunction<String>() {
//            @Override
//            public void call(String s) throws Exception {
//                System.out.println(s);
//            }
//        });

        //随机抽样算子
//        JavaRDD<String> sample = lines.sample(true, 0.1, 100L);
//        sample.foreach(new VoidFunction<String>() {
//            @Override
//            public void call(String s) throws Exception {
//                System.out.println(s);
//            }
//        });

        //action算子
//        List<String> collect = lines.collect();
//        for(String s:collect){
//            System.out.println(s);
//        }
//        String first = lines.first();
//        System.out.println(first);
//        List<String> take = lines.take(5);
//        for(String s: take){
//            System.out.println(s);
//        }
//        long count = lines.count();
//        System.out.println(count);



        //简化
//        JavaRDD<String> words = lines.flatMap(line -> Arrays.asList(line.split(" ")).iterator());
//        JavaPairRDD<String, Integer> pairWords = words.mapToPair(word -> new Tuple2<String, Integer>(word, 1));
//        JavaPairRDD<String, Integer> result = pairWords.reduceByKey((v1,v2) -> v1+v2);
//        result.foreach(tuple2 -> System.out.println(tuple2));

        //正常
//        JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
//            @Override
//            public Iterator<String> call(String line) throws Exception {
//                String[] split = line.split(" ");
//                return Arrays.asList(split).iterator();
//            }
//        });
//
//        JavaPairRDD<String,Integer> pairWords = words.mapToPair(new PairFunction<String, String, Integer>() {
//            @Override
//            public Tuple2<String, Integer> call(String word) throws Exception {
//                return new Tuple2<>(word,1);
//            }
//        });
//
//        JavaPairRDD<String,Integer> reduceResult = pairWords.reduceByKey(new Function2<Integer, Integer, Integer>() {
//            @Override
//            public Integer call(Integer v1, Integer v2) throws Exception {
//                return v1 + v2;
//            }
//        });
//
//        JavaPairRDD<Integer, String> swap = reduceResult.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
//
//            @Override
//            public Tuple2<Integer, String> call(Tuple2<String, Integer> tuple2) throws Exception {
//                return tuple2.swap();
//            }
//        });
//
//        JavaPairRDD<Integer,String> result = swap.sortByKey(false);
//
//        JavaPairRDD<String, Integer> end = result.mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
//
//            @Override
//            public Tuple2<String, Integer> call(Tuple2<Integer, String> integerStringTuple2) throws Exception {
//                return integerStringTuple2.swap();
//            }
//        });
//
//        end.foreach(new VoidFunction<Tuple2<String,Integer>>() {
//            @Override
//            public void call(Tuple2<String,Integer> Tuple2) throws Exception {
//                System.out.println(Tuple2);
//            }
//        });
        sc.stop();
    }
}

File: src/main/scala/corecode/WordCount.scala
package sqlcode

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCount {

    def main(args: Array[String]): Unit = {
      val conf = new SparkConf()
      conf.setMaster("local")
      conf.setAppName("scala-wc")
      val sc = new SparkContext(conf)
      val lines = sc.textFile("data/words")

      //算子
//      val result: RDD[String] = lines.filter(line => "hello hbase".equals(line))
//      result.foreach(println)
//      val result: RDD[String] = lines.sample(true,0.1,100L)
//      result.foreach(println)
//      val strings = lines.take(10)
//      var str = lines.first()
//      str.foreach(println)
//      val unit = sc.makeRDD(Array[Int](1,2,3,4,5,6))
//      unit.foreach(println)
      //正常
      val words: RDD[String] = lines.flatMap(line => {line.split(" ")})
      val pairWords = words.map(word=>{new Tuple2(word,1)})
      val reduceResult =pairWords.reduceByKey((v1:Int,v2:Int)=>{v1+v2})
      val result = reduceResult.sortBy(tp => {tp._2},false)
      //sortByKey
////      val result = reduceResult.map(tp =>{tp.swap}).sortByKey(false).map(tp => {tp.swap})
////      result.foreach(println)
//
//      //简化
////      val conf = new SparkConf().setAppName("scalawordcount").setMaster("local")
////      val sc = new SparkContext(conf)
////      sc.textFile("./data/words").flatMap(_.split(" ")).map((_, 1)).reduceByKey(_+_).sortBy(_._2,false).foreach(println)
  }
}
